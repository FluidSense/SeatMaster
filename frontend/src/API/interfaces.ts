// TODO: make the form component reflect Room type and change preferredRoom to IRoom
export interface IPostApplicationForm {
  comments?: string;
  seatRollover?: boolean;
  needs?: string;
  partnerUsername?: string;
  preferredRoom?: string;
}

export interface IPostAdminApplicationForm extends IPostApplicationForm {
  masterStatus?: string;
  status?: string;
}

export interface IPostApplicationSeason {
  newPeriodEnd: string;
  newPeriodStart: string;
  newRoomEnd: string;
  newRoomStart: string;
}

export interface IPutUserOnSeat {
  userId: number;
  roomId: number;
  seatId: string;
}

export interface IRemoveStudentFromSeat {
  roomId: number;
  seatId: string;
}

export interface IPostRoom {
  name: string;
  info: string;
}

export interface IRemoveStudentFromSeat {
  roomId: number;
  seatId: string;
}

// FIXME Move this to a fitting place and export the hell out of it
export interface IUser {
  email: string;
  fullname: string;
  id: number;
  username: string;
  admin: boolean;
}
