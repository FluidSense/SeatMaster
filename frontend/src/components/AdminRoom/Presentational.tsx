import AlertStripe from 'nav-frontend-alertstriper';
import HovedKnapp from 'nav-frontend-knapper';
import { Input, Textarea } from 'nav-frontend-skjema';
import { Sidetittel } from 'nav-frontend-typografi';
import React, { ChangeEvent, SyntheticEvent } from 'react';
import { ETIKETT_WARNING } from '../commonConstants';
import {
  _ALERT_CREATED_MESSAGE,
  _BUTTON_CREATE_ROOM,
  _BUTTON_DELETE_ROOM,
  _BUTTON_UPDATE_ROOM,
  _INPUT_LABEL_NAME,
  _INPUT_LABEL_NOTES,
  _TITLE_CREATE_NEW_ROOM,
  _TITLE_UPDATE_NEW_ROOM,
} from './strings';

interface IProps {
  roomName: string;
  roomNotes: string;
  setNotes: (roomNotes: ChangeEvent<HTMLInputElement>) => void;
  setName: (roomName: ChangeEvent<HTMLInputElement>) => void;
  buttonDisabled: boolean;
  onClick: () => void;
  deleteRoom: () => void;
  showAlert: boolean;
  alertMessage?: string;
  roomExists: boolean;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const {
    roomName,
    roomNotes,
    roomExists,
    setNotes,
    setName,
    buttonDisabled,
    onClick,
    deleteRoom,
    showAlert,
    alertMessage,
  } = props;
  const displayAlert =
    showAlert
      ? (
        <AlertStripe solid={true} type={ETIKETT_WARNING}>
          {alertMessage}
        </AlertStripe>)
      : null;
  const titleText = roomExists ? _TITLE_UPDATE_NEW_ROOM : _TITLE_CREATE_NEW_ROOM;
  const buttonText = roomExists ? _BUTTON_UPDATE_ROOM : _BUTTON_CREATE_ROOM;
  const deleteButton =
    roomExists
      ? (
        <HovedKnapp id={'delete-room-button'} type="fare" onClick={deleteRoom}>
          {_BUTTON_DELETE_ROOM}
        </HovedKnapp>)
      : null;
  // TextArea returns the wrong type, so its type has to be forced
  const assertEventType = (event: SyntheticEvent<EventTarget, Event>) => {
    const changeEvent = event as ChangeEvent<HTMLInputElement>;
    setNotes(changeEvent);
  };

  return (
    <>
      <Sidetittel>{titleText}</Sidetittel>
      {displayAlert}
      <Input
        id={'input-room-name'}
        onChange={setName}
        value={roomName}
        label={_INPUT_LABEL_NAME}
        bredde={'XL'}
      />
      <Textarea
        id={'input-room-notes'}
        onChange={assertEventType}
        value={roomNotes}
        label={_INPUT_LABEL_NOTES}
      />
      <div id="state-buttons">
        <HovedKnapp
          id={'create-room-button'}
          type="hoved"
          disabled={buttonDisabled}
          onClick={onClick}
        >
          {buttonText}
        </HovedKnapp>
        {deleteButton}
      </div>
    </>
  );
};

export default Presentational;
