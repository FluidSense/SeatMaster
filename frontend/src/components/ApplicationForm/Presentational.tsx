import AlertStripe from 'nav-frontend-alertstriper';
import KnappBase from 'nav-frontend-knapper';
import * as React from 'react';
import { postApplicationForm } from '../../API/calls';
import { IRegisteredUserState } from '../RegisterUser/reducer';
import { IRoom } from '../ViewRooms';
import ApplicationFormComments from './ApplicationFormComments';
import ApplicationFormPersonal from './ApplicationFormPersonal';
import ApplicationFormPreferences from './ApplicationFormPreferences';

import { _ALERT_USER_ERROR } from './Strings';

interface IProps {
  userInformation: IRegisteredUserState;
  rooms: IRoom[];
  changeModal: (modalOpen: boolean) => void;
  getRooms: () => void;
}

interface IState {
  room?: IRoom;
  partner: boolean;
  partnerUsername: string;
  needs: boolean;
  needsText: string;
  infoText: string;
  keepSeat: boolean;
  loading: boolean;
  error: string;
  [key: string]: string | boolean | IRoom | undefined;
}

export class Presentational extends React.Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      error: '',
      infoText: '',
      keepSeat: false,
      loading: false,
      needs: false,
      needsText: '',
      partner: false,
      partnerUsername: '',
      room: undefined,
    };
  }

  public componentDidMount() {
    this.props.getRooms();
  }

  public render() {
    const { userInformation } = this.props;
    const alertBox = this.state.error ? this.alertUser(_ALERT_USER_ERROR) : undefined;
    return (
      <>
        {alertBox}
        <form
          onSubmit={this.onSubmitForm}
          id="new-application-form"
        >
          <ApplicationFormPersonal
            fullname={userInformation.fullname}
            email={userInformation.email}
            status={userInformation.masterStatus}
          />
          <ApplicationFormPreferences
            updateApplicationFormData={this.updateApplicationFormData}
            partner={this.state.partner}
            rooms={this.props.rooms}
          />
          <ApplicationFormComments
            updateApplicationFormData={this.updateApplicationFormData}
            needs={this.state.needs}
          />
          <KnappBase
            id="submit-application"
            type="hoved"
            htmlType="submit"
            autoDisableVedSpinner={true}
            spinner={this.state.loading}
          >
            Submit
          </KnappBase>
        </form>
      </>
    );
  }

  private updateApplicationFormData = (item: React.FormEvent) => {
    const eventTarget = item.target as HTMLFormElement;
    const name: string = eventTarget.name;
    const value: string | boolean =
      eventTarget.type === 'checkbox' ? eventTarget.checked : eventTarget.value;
    this.setState({ [name]: value });
  }

  private onSubmitForm = (e: React.FormEvent) => {
    e.preventDefault();
    this.setState({ loading: true });

    postApplicationForm({
      comments: this.state.infoText,
      needs: this.state.needs ? this.state.needsText : '',
      partnerUsername: this.state.partnerUsername,
      preferredRoom: this.state.room ? this.state.room.name : this.state.room,
      seatRollover: this.state.keepSeat,
    })
      .then(
        // On fullfilled promise:
        () => {
          this.props.changeModal(true);
          this.setState({ loading: false, error: '' });
        },
        // On rejected promise:
        () => {
          this.setState({ loading: false, error: _ALERT_USER_ERROR });
        });
  }

  private alertUser = (text: string) => (
    <AlertStripe
      type="advarsel"
      solid={true}
    >
      {text}
    </AlertStripe>
  )
}

export default Presentational;
