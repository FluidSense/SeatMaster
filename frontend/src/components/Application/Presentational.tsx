import * as React from 'react';
import ApplicationForm from '../ApplicationForm';
import ConfirmationModal from './../ApplicationForm/ConfirmationModal';

interface IProps {
  modalIsOpen: boolean;
  changeModal: (modalOpen: boolean) => void;
}

export const Presentational: React.FunctionComponent<IProps> = (props) => {
  return (
    <>
      <ConfirmationModal modalIsOpen={props.modalIsOpen} changeModal={props.changeModal} />
      <ApplicationForm changeModal={props.changeModal} />
    </>
  );
};

export default Presentational;
