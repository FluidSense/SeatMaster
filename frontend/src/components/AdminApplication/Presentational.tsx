import React from 'react';
import { IPostAdminApplicationForm } from '../../API/interfaces';
import { IApplication } from '../Application';
import ApplicationOverview from '../ApplicationReview/ApplicationOverview';
import AssignSeat from '../AssignSeat';
import { IRoom } from '../ViewRooms';
import AcceptApplication from './AcceptApplication';
import './adminapplication.css';
import ApplicationSeatDisplay from './ApplicationSeatDisplay';
import SetWaitingList from './SetWaitingList';

interface IProps {
  rooms: IRoom[];
  application: IApplication;
  removeStudentFromSeat: (roomId: number, seatId: string) => void;
  updateApplication: (id: number, app: IPostAdminApplicationForm) => void;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const { application, rooms, removeStudentFromSeat, updateApplication } = props;
  if (!(application && rooms)) return null;
  const givenSeat = application.seat;
  const givenRoomId = givenSeat ? givenSeat.roomId : 0;
  const selectedRooms = rooms.filter(obj => obj.id === givenRoomId);

  return (
    <div className="main-content">
      <ApplicationOverview
        application={application}
        title={application.user ? application.user.fullname : ''}
        pathToEdit={`/admin/applications/${application.id}/edit`}
      />
      <div className="edit-lower-row">
        <AssignSeat rooms={rooms} application={application} />
        <ApplicationSeatDisplay
          seat={application.seat}
          room={selectedRooms[0]}
          removeFromSeat={removeStudentFromSeat}
        />
      </div>
      <div className="center-div">
        <SetWaitingList
          seat={application.seat}
          updateApplication={updateApplication}
          application={application}
        />
        <AcceptApplication application={application} updateApplication={updateApplication}/>
      </div>
    </div>
  );
};

export default Presentational;
