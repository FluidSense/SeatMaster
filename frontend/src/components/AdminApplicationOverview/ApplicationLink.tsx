import { LenkepanelBase } from 'nav-frontend-lenkepanel';
import { Element } from 'nav-frontend-typografi';
import React from 'react';
import { NavLink } from 'react-router-dom';
import { IApplication } from '../Application';
import { APP_APPROVED, APP_DENIED, APP_NOT_FOUND, APP_SUBMITTED } from '../commonConstants';
import { IRoom, ISeat } from '../ViewRooms';
import { ROUTE_TO } from './constants';
import {
  _LINK_APP_STATUS,
  _LINK_MASTER_STATUS,
  _LINK_NAME_TITLE,
  _LINK_ROOM_STATUS,
  _LINK_SEAT_STATUS,
} from './strings';

interface IProps {
  application: IApplication;
  rooms: IRoom[];
}

const getRoomNameFromId = (rooms: IRoom[], seat: ISeat) => {
  const room = rooms.find(x => (x.id === seat.roomId));
  if (room) return room.name;
  return '';
};

const ApplicationLink: React.FunctionComponent<IProps> = (props) => {
  const { application, rooms } = props;
  const user = application.user;
  if (!user || !application) return null;
  const status = application.status;
  const { seat } = application;
  const roomName = seat ? getRoomNameFromId(rooms, seat) : '';
  const seatId = seat ? seat.id : '';
  const link = (panelProps: any) => (
    <NavLink
      to={{ pathname: panelProps.href, application: { ...application }, rooms: [...rooms] }}
      className="lenkepanel lenkepanel--border"
    >
      {panelProps.children}
    </NavLink>);
  return (
    <LenkepanelBase
      key={application.id}
      linkCreator={link}
      href={`${ROUTE_TO}${application.id}`}
    >
      <div className="lenkepanel-applications">
        <div className="link-name"><Element>{_LINK_NAME_TITLE}</Element>{user.username}</div>
        <div className="link-master">
          <Element>{_LINK_MASTER_STATUS}</Element>
          {application.rank}
        </div>
        <div className="link-status"><Element>{_LINK_APP_STATUS}</Element>{status}</div>
        <div className="link-room"><Element>{_LINK_ROOM_STATUS}</Element>{roomName}</div>
        <div className="link-seat"><Element>{_LINK_SEAT_STATUS}</Element>{seatId}</div>
      </div>
    </LenkepanelBase>
  );
};

export default ApplicationLink;
