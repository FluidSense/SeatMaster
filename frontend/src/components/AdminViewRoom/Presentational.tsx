import KnappBase from 'nav-frontend-knapper';
import { Sidetittel } from 'nav-frontend-typografi';
import React from 'react';
import { Redirect } from 'react-router';
import { IPostAdminApplicationForm, IUser } from '../../API/interfaces';
import CSVButton from '../AdminRoom/CSVButton';
import { IApplication } from '../Application';
import { ISeat } from '../Seats';
import { IRoom } from '../ViewRooms';
import SeatPlacer from './SeatPlacer';
import { _EDIT_ROOM_BUTTON } from './strings';

interface IStateProps {
  match: {
    params: {
      id: string,
    },
  };
  applications?: IApplication[];
  rooms: IRoom[];
}

interface IDispatchProps {
  assign: (user: IUser, seat: ISeat) => void;
  delete: (seat: ISeat) => void;
  getAllApplications: () => void;
  fetchRooms: () => void;
  fetchRoom: (roomId: number) => void;
  updateApplication: (id: number, app: IPostAdminApplicationForm) => void;

}

interface IState {
  redirect: boolean;
  fetchedRooms: boolean;
  fetchedApplications: boolean;
}

type Props = IStateProps & IDispatchProps;

class Presentational extends React.Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      fetchedApplications: false,
      fetchedRooms: false,
      redirect: false,
    };
  }

  public componentDidMount() {
    if (!this.props.rooms.length && !this.state.fetchedRooms) {
      this.setState({ fetchedRooms: true });
      this.props.fetchRooms();
    }
  }

  public render() {
    const { applications, rooms, fetchRoom } = this.props;
    const roomId = parseInt(this.props.match.params.id, 10);
    const { redirect } = this.state;
    const room = rooms.find(iterroom => iterroom.id === roomId);
    if (!room) return null; // Insert spinner here
    const seats = this.seatsPlacers(room.seats.seats, applications);
    if (redirect) {
      return (
        <Redirect
          to={{ room, pathname: `/admin/rooms/${room.id}/update-room` }}
        />);
    }

    return (
      <div className="main-content">
        <div className="title-and-button">
          <Sidetittel>{room.name}</Sidetittel>
          <CSVButton room={room} fetchRoomInfo={fetchRoom} />
          <KnappBase type="hoved" onClick={this.setRedirect}>
            {_EDIT_ROOM_BUTTON}
          </KnappBase>
        </div>
        <h2>Info</h2>
        <p>{room.info}</p>
        <div>
          {seats}
        </div>
      </div>
    );
  }

  private setRedirect = () => {
    this.setState({ redirect: true });
  }

  private seatsPlacers = (seats: ISeat[], applications?: IApplication[]) => {
    if (!(applications && applications.length) && !this.state.fetchedApplications) {
      this.setState({ fetchedApplications: true });
      this.props.getAllApplications();
      return;
    }
    const apps = applications ? applications : [];
    const users = apps.map(application => application.user);
    return seats.map((seat, index) => {
      return (
        <SeatPlacer
          delete={this.props.delete}
          key={index}
          seat={seat}
          users={users}
          assign={this.props.assign}
          applications={this.props.applications}
          updateApplication={this.props.updateApplication}
        />
      );
    });
  }
}

export default Presentational;
