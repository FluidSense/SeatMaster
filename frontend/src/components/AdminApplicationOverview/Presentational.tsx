import { Sidetittel } from 'nav-frontend-typografi';
import React, { FunctionComponent } from 'react';
import { IApplication } from '../Application';
import { TitleAndSpinner } from '../LoadingPageSpinner/TitleAndSpinner';
import SearchBar, { searchBarEvent } from '../SearchBar';
import { IRoom } from '../ViewRooms';
import ApplicationLink from './ApplicationLink';
import ApproveSeated from './ApproveSeated';
import { FETCHING_APPLICATION_DATA } from './constants';
import PutOnWaitingList from './PutOnWaitingList';
import { _APPLICATIONS_TITLE } from './strings';

interface IProps {
  applications: IApplication[];
  rooms: IRoom[];
  fetching: string;
  approve: (ids: number[]) => void;
  filterFunction: (event: searchBarEvent) => void;
  putWaiting: (ids: number[]) => void;
}
const title = <Sidetittel>{_APPLICATIONS_TITLE}</Sidetittel>;
const emptyPage = (filterFunction: (event: searchBarEvent) => void) => (
  <div className="main-content">
    {title}
    <SearchBar filterFunction={filterFunction}/>
</div>);
const emptyPageWithLoader = (<TitleAndSpinner title={_APPLICATIONS_TITLE}/>);

const Presentational: FunctionComponent<IProps> = (props) => {
  const { applications, rooms, fetching, filterFunction, approve, putWaiting } = props;
  if (fetching === FETCHING_APPLICATION_DATA) return emptyPageWithLoader;
  if (!applications || !applications.length) return emptyPage(filterFunction);
  const applicationList = applications.map(application => (
    <ApplicationLink key={application.id} application={application} rooms={rooms} />
    ));
  return (
    <div className="main-content">
      <div className="title-and-button">
        <Sidetittel>
          {_APPLICATIONS_TITLE}
        </Sidetittel>
        <PutOnWaitingList putWaiting={putWaiting} applications={applications}/>
        <ApproveSeated approve={approve} applications={applications}/>
      </div>
      <SearchBar filterFunction={filterFunction}/>
      {applicationList}
    </div>
  );
};

export default Presentational;
