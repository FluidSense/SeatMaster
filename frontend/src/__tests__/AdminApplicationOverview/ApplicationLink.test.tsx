import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { IApplication } from '../../API/interfaces';
import ApplicationLink from '../../components/AdminApplicationOverview/ApplicationLink';
import { mockApplication } from './Presentational.test';

describe('Renders roomlink correctly', () => {
  const app: IApplication = mockApplication(1);
  it('Renders a room link correctly', () => {
    const wrapper = shallow(<ApplicationLink application={app} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
