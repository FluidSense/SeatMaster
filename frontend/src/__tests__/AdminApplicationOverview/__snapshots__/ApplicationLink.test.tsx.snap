// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders roomlink correctly Renders a room link correctly 1`] = `
<LenkepanelBase
  border={false}
  href="/admin/applications/1"
  key="1"
  linkCreator={[Function]}
>
  <div
    className="application-link"
  >
    <div
      className="link-name"
    >
      <Element>
        Name
      </Element>
      Kek
    </div>
    <div
      className="link-master"
    >
      <Element>
        Master Status
      </Element>
      Status
    </div>
    <div
      className="link-status"
    >
      <Element>
        Application Status
      </Element>
      Waiting list
    </div>
  </div>
</LenkepanelBase>
`;
